src/os/bluestore/HybridAllocator_impl.h

```diff
@@ -0,0 +1,186 @@
+  uint64_t max_alloc_size,
+  int64_t  hint,
+  PExtentVector* extents)
+{
+  dout(10) << __func__ << std::hex
+    << " 0x" << want
+    << "/" << unit
+    << "," << max_alloc_size
+    << "," << hint
+    << std::dec << dendl;
+  ceph_assert(std::has_single_bit(unit));
+  ceph_assert(want % unit == 0);
+
+  if (max_alloc_size == 0) {
+    max_alloc_size = want;
+  }
+  if (constexpr auto cap = std::numeric_limits<uint32_t>::max();
+    max_alloc_size >= cap) {
+    max_alloc_size = p2align(uint64_t(cap), (uint64_t)T::get_block_size());
+  }
+
+  std::lock_guard l(T::get_lock());
+
+  // try bitmap first to avoid unneeded contiguous extents split if
+  // desired amount is less than shortes range in AVL
+  bool primary_first = !(bmap_alloc &&
+                         bmap_alloc->get_free() &&
+                         want < T::_lowest_size_available());
+
+  int64_t res = __allocate_or_rollback(primary_first,
+    want, unit, max_alloc_size, hint, extents);
+  ceph_assert(res >= 0); // all the errors to be handled inside
+                         // __allocate_or_rollback()
+  if ((uint64_t)res < want) {
+    auto orig_size = extents->size();
+    int64_t res2 = 0;
+    // try alternate allocator
+    if (!primary_first) {
+      res2 = T::_allocate(want - res, unit, max_alloc_size, hint, extents);
+    } else if (bmap_alloc) {
+      res2 =
+        bmap_alloc->allocate(want - res, unit, max_alloc_size, hint, extents);
+    }
+    if (res2 >= 0) {
+      res += res2;
+    } else {
+      // allocator shouldn't return new extents on error
+      ceph_assert(orig_size == extents->size());
+    }
+  }
+  return res ? res : -ENOSPC;
+}
+
+template <typename T>
+void HybridAllocatorBase<T>::dump()
+{
+  std::lock_guard l(T::get_lock());
+  T::_dump();
+  if (bmap_alloc) {
+    bmap_alloc->dump();
+  }
+  dout(0) << __func__
+    << " avl_free: " << T::_get_free()
+    << " bmap_free: " << (bmap_alloc ? bmap_alloc->get_free() : 0)
+    << dendl;
+}
+
+template <typename T>
+void HybridAllocatorBase<T>::init_rm_free(uint64_t offset, uint64_t length)
+{
+  if (!length)
+    return;
+  std::lock_guard l(T::get_lock());
+  dout(10) << __func__ << std::hex
+    << " offset 0x" << offset
+    << " length 0x" << length
+    << std::dec << dendl;
+  T::_try_remove_from_tree(offset, length,
+    [&](uint64_t o, uint64_t l, bool found) {
+      if (!found) {
+        if (bmap_alloc) {
+          bmap_alloc->init_rm_free(o, l);
+        } else {
+          derr << __func__ << " lambda " << std::hex
+            << "Uexpected extent: "
+            << " 0x" << o << "~" << l
+            << std::dec << dendl;
+          ceph_assert(false);
+        }
+      }
+    });
+}
+
+template <typename T>
+void HybridAllocatorBase<T>::_spillover_range(uint64_t start, uint64_t end)
+{
+  auto size = end - start;
+  dout(20) << __func__
+    << std::hex << " "
+    << start << "~" << size
+    << std::dec
+    << dendl;
+  ceph_assert(size);
+  if (!bmap_alloc) {
```

Comment:
The main, if not only goal of hybrid was to save memory.

It is suboptimal that when we are consuming too much memory with ranged allocator, we create bitmap and consume even more.
We should have a strategy, that if we pass over some threshold, we create bitmap, but then move ranges to bitmap until we are again under memory limit.

Reply:
I'd rather rephrase the primary goal as "to avoid severe RAM usage growth". In that regard hybrid allocator prevents(!) that further growth by attaching additional bitmap allocator. Which has its RAM cost indeed.

So hybrid allocator's threshold is not a strict RAM bound but just a trigger to extend allocation policy.

One of the benefit of the current approach over the full switch to bitmap is that we keep effective using avl allocator for tracking/allocating long contiguous extents. Which is a known issue for bitmap allocator. While bitmap counterpart to be used for tracking small single-au extents - which actually tend to cause high RAM consumption in AVL

So in other words - this hybrid scenario doesn't unconditionally save allocator's memory  to the limit but rather provides a trade-off between RAM usage and allocation speed.
Not to mention that this mechanics isn't new, it's borrowed from existing hybrid allocator...